astropy/io/fits/hdu/table.py

```diff
@@ -872,24 +872,7 @@ def _calculate_datasum_with_heap(self):
             # Now add in the heap data to the checksum (we can skip any gap
             # between the table and the heap since it's all zeros and doesn't
             # contribute to the checksum
-            if data._get_raw_data() is None:
-                # This block is still needed because
-                # test_variable_length_table_data leads to ._get_raw_data
-                # returning None which means _get_heap_data doesn't work.
-                # Which happens when the data is loaded in memory rather than
-                # being unloaded on disk
-                for idx in range(data._nfields):
-                    if isinstance(data.columns._recformats[idx], _FormatP):
-                        for coldata in data.field(idx):
-                            # coldata should already be byteswapped from the call
-                            # to _binary_table_byte_swap
-                            if not len(coldata):
-                                continue
-
-                            csum = self._compute_checksum(coldata, csum)
-            else:
-                csum = self._compute_checksum(data._get_heap_data(), csum)
-
+            csum = self._compute_checksum(data._get_heap_data(), csum)
```

Comment:
can just return this directly (another nitpick... but `ruff` will get you otherwise...)

Reply:
I have a slight preference for the current formulation because it makes it more explicit that the checksum is both the input and output of this function.